website/project/model.py

```diff
@@ -215,84 +236,157 @@ def get_content(self, auth):
         return self.content
 
     @classmethod
-    def find_unread(cls, user, node):
+    def find_unread(cls, user, node, page=None, root_id=None, check=False):
         default_timestamp = datetime.datetime(1970, 1, 1, 12, 0, 0)
         n_unread = 0
         if node.is_contributor(user):
-            view_timestamp = user.comments_viewed_timestamp.get(node._id, default_timestamp)
-            n_unread = Comment.find(Q('node', 'eq', node) &
-                                    Q('user', 'ne', user) &
-                                    Q('date_created', 'gt', view_timestamp) &
-                                    Q('date_modified', 'gt', view_timestamp)).count()
+            if not page:
+                return cls.n_unread_node_comments(user, node) + cls.n_unread_file_comments(user, node)
+            elif page == 'node':
+                return cls.n_unread_node_comments(user, node)
+            elif page == 'files':
+                if root_id is None:
+                    return cls.n_unread_file_comments(user, node)
+                else:
+                    view_timestamp = user.get_node_comment_timestamps(node, page)
+                    if isinstance(view_timestamp, dict):
+                        view_timestamp = view_timestamp.get(root_id, default_timestamp)
+                    root_target = File.load(root_id)
+                    if check:
+                        if not (root_target and root_target.touch(request.headers.get('Authorization'))):
+                            return 0
+                    else:
+                        return Comment.find(Q('node', 'eq', node) &
+                                            Q('user', 'ne', user) &
+                                            Q('is_deleted', 'eq', False) &
+                                            (Q('date_created', 'gt', view_timestamp) |
+                                            Q('date_modified', 'gt', view_timestamp)) &
+                                            Q('root_target', 'eq', root_target)).count()
+
+        return n_unread
+
+    @classmethod
+    def n_unread_node_comments(cls, user, node):
+        view_timestamp = user.get_node_comment_timestamps(node, 'node')
+        return Comment.find(Q('node', 'eq', node) &
+                            Q('user', 'ne', user) &
+                            (Q('date_created', 'gt', view_timestamp) |
+                            Q('date_modified', 'gt', view_timestamp)) &
+                            Q('is_deleted', 'eq', False) &
+                            Q('root_target', 'eq', node)).count()
+
+    @classmethod
+    def n_unread_file_comments(cls, user, node):
+        n_unread = 0
+        for file_id in node.commented_files:
+            file_obj = File.load(file_id)
```

Comment:
Would it be better to do a `File.find(Q('_id', 'in', node.commented_files))` and then loop through that so that you're limiting the mongo queries?